core/trino-main/src/main/java/io/trino/node/AnnounceNodeAnnouncer.java

```diff
@@ -0,0 +1,182 @@
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static io.airlift.http.client.Request.Builder.preparePost;
+import static io.airlift.http.client.StaticBodyGenerator.createStaticBodyGenerator;
+import static io.airlift.http.client.StatusResponseHandler.createStatusResponseHandler;
+import static jakarta.ws.rs.core.MediaType.TEXT_PLAIN;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class AnnounceNodeAnnouncer
+        implements Announcer
+{
+    private static final Logger log = Logger.get(AnnounceNodeAnnouncer.class);
+
+    private final HttpClient httpClient;
+    private final List<URI> announceUris;
+    private final ScheduledExecutorService announceExecutor;
+    private final StaticBodyGenerator currentHostAnnouncement;
+    private final AtomicLong lastWarningLogged = new AtomicLong(0);
+
+    private final AtomicReference<State> state = new AtomicReference<>(State.NOT_STARTED);
+
+    private enum State
+    {
+        NOT_STARTED,
+        STARTED,
+        STOPPED,
+    }
+
+    @Inject
+    public AnnounceNodeAnnouncer(InternalNode currentNode, AnnounceConfig config, @ForAnnouncer HttpClient httpClient)
+    {
+        this(currentNode.getInternalUri(), config.getCoordinatorUris(), httpClient);
+    }
+
+    @VisibleForTesting
+    public AnnounceNodeAnnouncer(URI internalUri, Collection<URI> coordinatorUris, HttpClient httpClient)
+    {
+        URI currentUri = internalUri;
+        this.announceUris = coordinatorUris.stream()
+                .filter(uri -> !currentUri.equals(uri))
+                .map(uri -> uriBuilderFrom(uri).appendPath("/v1/announce").build())
+                .distinct()
+                .toList();
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+        this.announceExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("node-state-poller-%s"));
+        this.currentHostAnnouncement = createStaticBodyGenerator(currentUri.toString(), UTF_8);
+    }
+
+    @Override
+    public void start()
+    {
+        if (announceUris.isEmpty()) {
+            log.warn("No coordinator URIs configured, skipping node state announcements");
+            return;
+        }
+
+        checkState(!announceExecutor.isShutdown(), "Announcer has been destroyed");
+        if (!state.compareAndSet(State.NOT_STARTED, State.STARTED)) {
+            log.warn("Announcer is already started, skipping redundant start");
+            return;
+        }
+
+        announceExecutor.scheduleWithFixedDelay(
+                () -> {
+                    if (state.get() != State.STARTED) {
```

Comment:
Why is this needed? At this point, it can't be in `NOT_STARTED`, and if it was `STOPPED`, then this can only happen in a tiny window (and doesn't seem to matter)